### PR TITLE
2nd PR: Replace Apache HttpClient with com.cloudant.http.Http class

### PR DESCRIPTION
**Pull request by emlaver**
*What*
This is the second part for replacing the http dependency from Apache’s httpclient to Java’s HttpURLConnection.  Migrating to HttpUrlConnection is required for us to use java-cloudant within sync-android.

*How*
- Replace HttpPut and RESTful commands of Apache’s httpclient with sync-android’s Http (using Java’s HttpUrlConnection)
- Replace HttpResponse with HttpConnection’s InputStream response
- Remove imports for org.apache.http
- Remove any methods that were only specific to Apache’s httpclient (i.e. connection manager)
- Remove CouchDb Android class
- Remove shutdown method from CloudantClient and CouchDbClient
- Add URI’s getUserInfo to URIBuilder and CouchDbProperties.  This is needed for setting auth property in HttpConnection
- Use sync-android’s executeToInputStream for executing HTTP calls and properly handling the response
- Add Rich's commit for intermittent problems with Replicator and Replication tests (Rich’s branch 52593-replicate-tests).
- Remove shutdown comment in javadoc
- Fix return javadoc for HTTP API
- Remove shutdown method from test cases
- Ignore SSL test cases - will enable in next PR with Rich’s interceptor branch (53085-ssl-setters)
- Ignore cookieNegativeTest - enable in next PR with cookie interceptor fixes
- Ignore Replicator tests. Currently, they succeed for CouchDB but several fail for Cloudant database.       
  Will be enabled with the PR from 52593-replicate-tests.

*Testing*
Use existing test cases. Test cases from sync-android in next PR (i.e. HttpTest)

reviewer @ricellis 
reviewer @rhyshort

### CODE REVIEW

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 1]
```diff
@@ -236,7 +237,30 @@ public CloudantClient(String account, String authCookie, ConnectOptions connectO
...
+
+        doInit(h.get("scheme"),
+                h.get("hostname"),
+                new Integer(h.get("port")).intValue(),
+                h.get("username"),
+                h.get("password"),
+                null,
+                null, //authcookie
+                interceptor
+        );
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: I don't think we should have any API exposure of the `CookieInterceptor` here. We should also leave the constructor argument changes for the other ticket.

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 2]
```diff
@@ -247,7 +271,8 @@ public CloudantClient(String account, String authCookie, ConnectOptions connectO
      */
     public ApiKey generateApiKey() {
         URI uri = buildUri(getBaseUri()).path("_api/v2/api_keys").build();
-        return client.executeRequest(createPost(uri, "", ""), ApiKey.class);
+        HttpConnection connection = Http.POST(uri, "application/json");
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: We should keep the usage of `HttpConnection` in as few classes as possible. We will need to set interceptors on every created `HttpConnection` and also potentially other parameters (e.g. `Proxy`). As such it would be beneficial to have all usages in one place.
So a better solution here might be to call a `post(uri)` method (e.g. the one on `CouchDatabaseBase`).

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 3]
```diff
@@ -388,24 +403,19 @@ public String serverVersion() {
...
+        InputStream response = client.executeRequest(request);
         return response;
 
     }
 
-
-    /**
-     * Shuts down the connection manager used by this client instance.
-     */
-    public void shutdown() {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: I think we should leave the `shutdown()` method in place even if it is currently a no-op. It is already part of the API and it preserves it for future use.

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 4]
```diff
@@ -484,10 +503,10 @@ Response put(URI uri, Object object, boolean newEntity, int writeQuorum, Gson gs
                 assertNotEmpty(id, "id");
                 assertNotEmpty(rev, "rev");
             }
-            final HttpPut put = new HttpPut(buildUri(uri).pathToEncode(id).query("w",
-                    writeQuorum).buildEncoded());
-            CouchDbUtil.setEntity(put, json.toString(), "application/json");
-            response = client.executeRequest(put);
+            HttpConnection putConnection = Http.PUT(buildUri(uri).pathToEncode(id).query("w",
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Another instance where a `put()` method might be better. Although thinking about it more maybe we will be ok as we can fix up the `executeRequest` method to add the interceptors/proxy etc anyway.

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 5]
```diff
@@ -530,17 +553,31 @@ Response put(URI uri, Object object, boolean newEntity, int writeQuorum, Gson gs
      */
     private void doInit(String scheme, String hostname, int port,
                         String loginUsername, String password, ConnectOptions connectOptions,
-                        String authCookie) {
+                        String authCookie, CookieInterceptor interceptor) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Take this `CookieInterceptor` out of the constructor for now and create it from the credentials in this method. Obviously we'll need to expose setters for the interceptors later.

File: src/main/java/com/cloudant/client/api/Database.java [hunk 6]
```diff
@@ -211,10 +207,10 @@ private JsonObject getPermissionsObject() {
      * @return List of shards
      */
     public List<Shard> getShards() {
-        HttpResponse response = null;
-        HttpGet get = new HttpGet(buildUri(db.getDBUri()).path("/_shards").build());
+        InputStream response = null;
+        //HttpGet get = new HttpGet(buildUri(db.getDBUri()).path("/_shards").build());
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Database.java]: Remove commented out code.

File: src/main/java/com/cloudant/client/api/Database.java [hunk 7]
```diff
@@ -211,10 +207,10 @@ private JsonObject getPermissionsObject() {
...
         try {
-            response = client.executeRequest(get);
+            response = client.get(buildUri(db.getDBUri()).path("/_shards").build());
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Database.java]: Does this `/` before `_shards` need removing? In other places they seem to have been.

File: src/main/java/com/cloudant/client/api/Database.java [hunk 8]
```diff
@@ -258,10 +254,10 @@ public void createIndex(String indexName, String designDocName, String indexType
      */
     public void createIndex(String indexDefinition) {
         assertNotEmpty(indexDefinition, "indexDefinition");
-        HttpResponse putresp = null;
+        InputStream putresp = null;
         URI uri = buildUri(getDBUri()).path("_index").build();
         try {
-            putresp = client.executeRequest(createPost(uri, indexDefinition, "application/json"));
+            putresp = client.executeRequest(createPost(uri, null, "application/json"));
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Database.java]: Should this `null` be `indexDefinition`?

File: src/main/java/com/cloudant/client/api/Database.java [hunk 9]
```diff
@@ -957,10 +953,14 @@ private String getFindByIndexBody(String selectorJson,
         return finalbody.toString();
     }
 
-    HttpResponse executeRequest(HttpRequestBase request) {
+    InputStream executeRequest(HttpConnection request) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Database.java]: This method should probably be removed, it seems like most places call `client.executeRequest()` anyway and if they don't they probably should.

File: src/main/java/com/cloudant/client/api/Database.java [hunk 10]
```diff
@@ -957,10 +953,14 @@ private String getFindByIndexBody(String selectorJson,
...
         return client.executeRequest(request);
     }
 
+    <T> T executeRequest(HttpConnection request, Class<T> classType) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Database.java]: Also should be removed.

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 11]
```diff
@@ -388,24 +403,19 @@ public String serverVersion() {
...
 
     }
 
-
-    /**
-     * Shuts down the connection manager used by this client instance.
-     */
-    public void shutdown() {
-        client.shutdown();
+    public <T> T executeRequest(HttpConnection request, Class<T> classType) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Why do we need this parameterized version? Isn't the `InputStream` version sufficient? If it isn't then shouldn't we just have the parameterized one (optionally being typed to `InputStream`).

File: src/main/java/com/cloudant/client/api/Search.java [hunk 12]
```diff
@@ -119,9 +119,9 @@
     public InputStream queryForStream(String query) {
         key(query);
         URI uri = uriBuilder.build();
-        HttpGet get = new HttpGet(uri);
-        get.addHeader("Accept", "application/json");
-        return getStream(db.executeRequest(get));
+        HttpConnection get = Http.GET(uri);
+        get.getConnection().addRequestProperty("Accept", "application/json");
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/Search.java]: I think the form `get.requestProperties.add(...)` is preferred.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java [hunk 13]
```diff
@@ -89,11 +56,8 @@
...
  * <p/>
  * <p/>
- * <p>At the end of a client usage; it's useful to call: {@link #shutdown()} to ensure proper
- * release of resources.
  *
  * @author Ahmed Yehia
- * @see CouchDbClientAndroid
  * @since 0.0.2
  */
 public class CouchDbClient extends CouchDbClientBase {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java]: As there is essentially nothing left here perhaps we should move the remaining `database()` method to `CouchDbClientBase` and then delete this class and rename `CouchDbClientBase` to `CouchDbClient`.
Wdyt?

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 14]
```diff
@@ -506,6 +525,10 @@ Response put(URI uri, Object object, boolean newEntity, int writeQuorum, Gson gs
                 h.put("scheme", uri.getScheme());
                 h.put("hostname", uri.getHost());
                 h.put("port", new Integer(uri.getPort()).toString());
+                if(uri.getUserInfo() != null && !uri.getUserInfo().isEmpty()) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Why is this added here? I thought this already happened in `HttpConnection`'s execute method.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 15]
```diff
@@ -77,85 +63,49 @@
...
+
         this.gson = GsonHelper.initGson(new GsonBuilder()).create();
         final String path = props.getPath() != null ? props.getPath() : "";
-        this.baseURI = buildUri().scheme(props.getProtocol()).host(props.getHost()).port(props
-                .getPort()).path("/").path(path).build();
-
-        // for the authentication using previous session's cookie . do not get the cookie
-        if (props.getAuthCookie() == null) {
-            getCookie(props);
+        //Add username and password if authentication info exists
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Is this temporary until we have the `CookieInterceptor` enabled?

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 16]
```diff
@@ -77,85 +63,49 @@
...
-    /**
-     * @return {@link HttpContext} instance for HTTP request execution.
-     */
-    abstract HttpContext createContext();
-
-    /**
-     * Shuts down the connection manager used by this client instance.
-     */
-    abstract void shutdown();
+    // Client(s) provided implementation
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: If  `database` method is moved into here this comment can be deleted.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 17]
```diff
@@ -307,20 +241,29 @@ public Replicator replicator() {
...
-     * <p><b>Note</b>: The response must be closed after use to release the connection.
+     * <p><b>Note</b>: The stream must be closed after use to release the connection.
      *
-     * @param request The HTTP request to execute.
-     * @return {@link HttpResponse}
+     * @param connection The HTTP request to execute.
+     * @return Class type of object T (i.e. {@link Response}
      */
-    public HttpResponse executeRequest(HttpRequestBase request) {
+    public <T> T executeRequest(HttpConnection connection, Class<T> classType) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Are the only acceptable types here `InputStream` and `Response`?

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 18]
```diff
@@ -330,64 +273,92 @@ public HttpResponse executeRequest(HttpRequestBase request) {
...
-     * @return {@link InputStream}
+     * @return Class type of object T (i.e. {@link Response}
      */
-    InputStream get(URI uri) {
-        HttpGet get = new HttpGet(uri);
-        get.addHeader("Accept", "application/json");
-        return get(get);
+    public <T> T get(URI uri, Class<T> classType) {
+        HttpConnection connection = Http.GET(uri);
+        if(classType.isInstance(HttpConnection.class)) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: What is going on here? This seems very strange as the get/put/post/delete methods would be expected to actually do a request, but this branch appears to convert a URI into a `HttpConnection`, so why not just do `Http.GET(uri)` in the caller?

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 19]
```diff
@@ -330,64 +273,92 @@ public HttpResponse executeRequest(HttpRequestBase request) {
...
+            return (T)connection;
+        } else {
+            //return executeToJsonObject(connection, classType);
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: remove commented out code

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 20]
```diff
@@ -330,64 +273,92 @@ public HttpResponse executeRequest(HttpRequestBase request) {
...
+        return executeRequest(connection);
+    }
+
+    /**
+     * Performs a HTTP PUT request, saves or updates a document.
+     *
+     * @return Input stream with response
+     */
+    InputStream put(URI uri) {
+        return put(uri, null);
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Improve javadoc to say this defaults to "application/json" content type and then pass "application/json" String, instead of `null`.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 21]
```diff
@@ -330,64 +273,92 @@ public HttpResponse executeRequest(HttpRequestBase request) {
...
+    /**
+     * Performs a HTTP PUT request with content type, saves or updates a document.
+     *
+     * @return Input stream with response
+     */
+    InputStream put(URI uri, String contentType) {
+        HttpConnection connection = null;
+        if(contentType != null) {
+            connection = Http.PUT(uri, contentType);
+        } else {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: remove this `if/else` based on https://github.com/cloudant/java-cloudant/pull/104/files?diff=split#r41374990

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 22]
```diff
@@ -330,64 +273,92 @@ public HttpResponse executeRequest(HttpRequestBase request) {
...
+     * @return {@link Response}
      */
-    HttpResponse head(URI uri) {
-        return executeRequest(new HttpHead(uri));
+    Response put(URI uri, InputStream instream, String contentType) {
+        HttpConnection connection;
+        try {
+            connection = Http.PUT(uri, contentType);
+
+            StringWriter writer = new StringWriter();
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Remove this `StringWriter` - `HttpConnection#setRequestBody(InputStream)` would be better.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 23]
```diff
@@ -512,17 +479,114 @@ public Gson getGson() {
...
      *
-     * @param request  The HTTP request to execute.
-     * @param Class<T> class of domain object to return
-     * @return <T> T
+     * @param connection  The HTTP connection to execute.
+     * @return <T> T json object
      */
-    public <T> T executeRequest(HttpRequestBase request, Class<T> classType) {
-        HttpResponse response = null;
+    public InputStream executeRequest(HttpConnection connection) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Why do we need this method? Can't the `executeToInputStream` on line 495 just be public?

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java [hunk 24]
```diff
@@ -512,17 +479,114 @@ public Gson getGson() {
...
+        return executeToInputStream(connection);
+    }
+
+    // - if 2xx then return stream
+    // - map 404 to NoResourceException
+    // - if there's a couch error returned as json, un-marshall and throw
+    // - anything else, just throw the IOException back, use the cause part of the exception?
+
+    // it needs to catch eg FileNotFoundException and rethrow to emulate the previous exception handling behaviour
+    private InputStream executeToInputStream(HttpConnection connection) throws CouchDbException {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: I think there would be some value to extracting this into a class of its own and potentially with a response object (of a new type) instead of `InputStream`, but I think that can happen later.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java [hunk 25]
```diff
@@ -229,4 +251,39 @@ public CouchDbProperties setAuthenticatedModeSSLSocketFactory(SSLSocketFactory f
         this.authenticatedModeSSLSocketFactory = factory;
         return this;
     }
+
+    public Map<String, String> getCustomHeaders() {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java]: custom headers don't need to be supported here (we'll just use interceptors if we need a custom header).

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java [hunk 26]
```diff
@@ -229,4 +251,39 @@ public CouchDbProperties setAuthenticatedModeSSLSocketFactory(SSLSocketFactory f
...
+        return customHeaders;
+    }
+
+    public CouchDbProperties setCustomHeaders(Map<String, String> customHeaders) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java]: custom headers don't need to be supported here either

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java [hunk 27]
```diff
@@ -211,11 +199,13 @@ public static String streamToString(InputStream in) {
     /**
      * Closes the response input stream.
      *
-     * @param response The {@link HttpResponse}
+     * @param response The {@link HttpConnection}
      */
-    public static void close(HttpResponse response) {
+    public static void close(InputStream response) {
         try {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: As IOUtils has been imported why not just use `IOUtils.closeQuietly()`?

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java [hunk 28]
```diff
@@ -227,7 +217,9 @@ public static void close(HttpResponse response) {
      */
     public static void close(Closeable c) {
         try {
-            c.close();
+            if(c != null) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: closeQuietly again?

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java [hunk 29]
```diff
@@ -257,36 +248,37 @@ public static String getAsString(HttpResponse response, String e) {
...
-     * @param response The {@link HttpResponse}
+     * @param response The response from {@link com.cloudant.http.HttpConnection}
+     * @param klazz
      * @return {@link Response}
      */
-    public static <T> List<T> getResponseList(HttpResponse response, Gson gson, Class<T> klazz,
+    public static <T> List<T> getResponseList(InputStream response, Gson gson, Class<T> klazz,
                                               Type typeofT) throws CouchDbException {
-        InputStream instream = getStream(response);
+        //InputStream instream = getStream(response);
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: Remove commented out code

File: src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java [hunk 30]
```diff
@@ -38,17 +38,26 @@
...
 
     public static URIBuilder buildUri() {
         return new URIBuilder();
     }
 
     public static URIBuilder buildUri(URI uri) {
-        URIBuilder builder = URIBuilder.buildUri().scheme(uri.getScheme())
-                .host(uri.getHost()).port(uri.getPort()).path(uri.getPath());
+        URIBuilder builder;
+        if(uri.getUserInfo() != null && !uri.getUserInfo().isEmpty()) {
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java]: I'm really confused as to why the user info stuff appears in so many places. I'm pretty sure this is going to go away when the `CookieInterceptor` is enabled though right?

File: src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java [hunk 31]
```diff
@@ -106,6 +127,11 @@ public URIBuilder path(String path) {
         return this;
     }
 
+    public URIBuilder userInfo(String userInfo) {
+        this.userInfo += userInfo;
```
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java]: Should this really be appending?
I feel like this should just do a set.

File: src/test/java/com/cloudant/tests/AttachmentsTest.java [hunk 32]
```diff
@@ -51,7 +51,6 @@ public void setUp() {
     @After
     public void tearDown() {
         account.deleteDB("lightcouch-db-test");
-        account.shutdown();
```
Comment (ricellis) [inline src/test/java/com/cloudant/tests/AttachmentsTest.java]: A bunch of these test changes won't be necessary if the `shutdown()` method remains in place.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 4 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: What if I move this put method to `CouchDbClientBase` class?

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 17 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Yes - currently the only return types are InputStream and Response.  Maybe it's better to just have two methods?

File: src/main/java/com/cloudant/client/api/Database.java (hunk 9 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Database.java]: `Search` class uses this in the `queryForStream` method and does not have a `client` object.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java (hunk 13 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java]: Yes, that's a good idea.  I wasn't completely clear of the purpose of this class and it seemed to add obscurity.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 24 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Yes, I agree.

File: src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java (hunk 30 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java]: The reason I added getUserInfo is to support the full URI with username:password in HttpConnection's execute.  If getUserInfo is null, it won't add the proper authentication and none of the HTTP requests will work:
https://github.com/cloudant/java-cloudant/blob/856de65030fdfbff6fb731b75daa4ceb77647d8b/src/main/java/com/cloudant/http/HttpConnection.java#L190

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 1 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 2 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 3 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 4 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Partial fix in e6dd315.  Left a note for tomorrow to discuss if we should move the method into `CouchDbClient` class.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 5 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/Database.java (hunk 6 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Database.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/Database.java (hunk 7 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Database.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/Database.java (hunk 8 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Database.java]: Added back. Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/Database.java (hunk 10 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Database.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 11 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/Search.java (hunk 12 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/Search.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java (hunk 13 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/api/CloudantClient.java (hunk 14 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 15 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Yes - I left a note on another userInfo comment pointing to HttpConnection's execute method that uses userInfo to create authentication.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 16 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 18 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 19 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 20 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 21 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 22 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java (hunk 23 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClientBase.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java (hunk 25 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java (hunk 26 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbProperties.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java (hunk 27 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java (hunk 28 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java (hunk 29 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/CouchDbUtil.java]: Fixed in e6dd315.

File: src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java (hunk 31 above)
Comment (emlaver) [inline src/main/java/com/cloudant/client/org/lightcouch/internal/URIBuilder.java]: I temporarily need a way to set the username and password into the URI.  If not, the proper authentication will not be added in HttpConnection's execute.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java (hunk 13 above)
Comment (ricellis) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java]: Yeah, I believe the structure was such that the Android and Java sub-classes of `CouchDbClientBase` could each provide some specific behaviour whilst sharing a bunch of code from the abstract base class. Hopefully we can keep all the Android specifics in the HTTP classes so we don't need the division.

File: src/main/java/com/cloudant/client/api/CloudantClient.java [hunk 33]
```diff
@@ -128,10 +127,6 @@
  * <h2>Replication</h2>
  * Replication {@link Replication account.replication()} and {@link Replicator account.replicator()}
  *
- * <h2>Shutting down</h2>
- * At the end of a client usage; it's useful to call {@link #shutdown() client.shutdown()} to
- * ensure proper release of resources.
- *
```
Comment (rhyshort) [inline src/main/java/com/cloudant/client/api/CloudantClient.java]: This should be reverted too, even though shutdown is now a No-op. Its better to have people still call so they are used to doing it in the event we need to make a change.

File: src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java [hunk 34]
```diff
@@ -1,333 +1,592 @@
...
+     * @param properties An object containing configuration properties.
+     * @see {@link CouchDbProperties}
+     */
+    public CouchDbClient(CouchDbProperties properties) {
+        this(new CouchDbConfig(properties));
+    }
+
+    /**
+     * Shuts down the connection manager used by this client instance.
+     */
```
Comment (rhyshort) [inline src/main/java/com/cloudant/client/org/lightcouch/CouchDbClient.java]: JavaDoc should be changed to reflect that we don't use a connection manager anymore

File: src/test/java/com/cloudant/tests/CloudantClientTests.java [hunk 35]
```diff
@@ -66,7 +67,6 @@ public void setUp() {
     @After
     public void tearDown() {
         account.shutdown();
-        cookieBasedClient.shutdown();
```
Comment (rhyshort) [inline src/test/java/com/cloudant/tests/CloudantClientTests.java]: I don't think this should be removed, since looking at the setup code the `cookieBasedClient` isn't the same as account.

File: src/test/java/com/cloudant/tests/SslAuthenticationTest.java [hunk 36]
```diff
@@ -162,9 +164,7 @@ public static void tearDownClass() {
 
     @After
     public void tearDown() {
-        if (dbClient != null) {
-            dbClient.shutdown();
-        }
+        dbClient.shutdown();
```
Comment (rhyshort) [inline src/test/java/com/cloudant/tests/SslAuthenticationTest.java]: do we still need the if null check?

File: src/test/java/com/cloudant/tests/ViewsTest.java [hunk 37]
```diff
@@ -68,7 +68,6 @@ public void setUp() {
     @After
     public void tearDown() {
         account.deleteDB("lightcouch-db-test");
-        account.shutdown();
```
Comment (rhyshort) [inline src/test/java/com/cloudant/tests/ViewsTest.java]: Should this have been reverted too?
